cpp/ql/src/semmlecode.cpp.dbscheme

```diff
@@ -408,7 +408,22 @@ function_deleted(unique int id: @function ref);
 
 function_defaulted(unique int id: @function ref);
 
+/**
+ * Holds if the function is declared to be `constexpr`.
+ */
+function_declared_constexpr(unique int id: @function ref);
 
+/**
+ * Holds if the function is `constexpr`. Normally, this holds if and
+ * only if `function_declared_constexpr` holds, but in some
+ * circumstances they differ. For example, with
+ * ```
+ * int f(int i) { return 6; }
+ * template <typename T> constexpr int g(T x) { return f(x); }
+ * ```
+ * `g<int>` is declared constexpr, but is not constexpr.
+ */
+function_is_constexpr(unique int id: @function ref);
```

Comment:
To close CPP-314 we'll also need to know which _variables_ are `constexpr`. See [the tests in #770](https://github.com/Semmle/ql/blob/a383a1d6c9e2288ccb171e58238a23e22ba49dbd/cpp/ql/test/library-tests/constants/addresses/addresses.cpp#L22-L61). To avoid adding more tables later, I suggest changing this to `declaration_is_constexpr(unique int id: @declaration ref)`.